src/rgw/rgw_dedup_store.cc

```diff
@@ -0,0 +1,689 @@
+                  const struct key_t       *p_key,
+                  const DoutPrefixProvider *dpp)
+  {
+    std::string oid(block_id.get_slab_name(md5_shard));
+    //p_obj_ioctx->set_namespace(itr->get_nspace());
+    int read_len = DISK_BLOCK_SIZE;
+    int byte_offset = block_id.get_block_offset() * DISK_BLOCK_SIZE;
+    bufferlist bl;
+    int ret = ioctx.read(oid, bl, read_len, byte_offset);
+    if (ret < 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::ERR: failed to read block from "
+                        << oid << ", error is " << cpp_strerror(-ret) << dendl;
+      return ret;
+    }
+    else {
+      ldpp_dout(dpp, 20) << __func__ << "::oid=" << oid << ", len=" << bl.length() << dendl;
+    }
+    const char *p = nullptr;
+    auto bl_itr = bl.cbegin();
+    size_t n = bl_itr.get_ptr_and_advance(sizeof(disk_block_t), &p);
+    if (n == sizeof(disk_block_t)) {
+      disk_block_t *p_disk_block = (disk_block_t*)p;
+      disk_block_header_t *p_header = p_disk_block->get_header();
+      p_header->deserialize();
+      ret = p_header->verify(block_id, dpp);
+      if (ret != 0) {
+        return ret;
+      }
+
+      unsigned offset = p_header->rec_offsets[rec_id];
+      // We deserialize the record inside the CTOR
+      disk_record_t rec(p + offset);
+      uint32_t size_4k_units = byte_size_to_disk_blocks(rec.s.obj_bytes_size);
+      struct key_t key(rec.s.md5_high, rec.s.md5_low, size_4k_units, rec.s.num_parts);
+      if (key == *p_key) {
+        *p_rec = rec;
+        return 0;
+      }
+      else {
+        ldpp_dout(dpp, 5) << __func__ << "::ERR: Bad record in block=" << block_id
+                          << ", rec_id=" << rec_id << dendl;
+        return -EIO;
+      }
+    }
+    else {
+      ldpp_dout(dpp, 5) << __func__ << "::ERR: unexpected short read n=" << n << dendl;
+      return -EIO;
+    }
+
+    return 0;
+  }
+
+  //---------------------------------------------------------------------------
+  int load_slab(librados::IoCtx &ioctx,
+                bufferlist &bl,
+                md5_shard_t md5_shard,
+                work_shard_t worker_id,
+                uint32_t seq_number,
+                const DoutPrefixProvider* dpp)
+  {
+    disk_block_id_t block_id(worker_id, seq_number);
+    std::string oid(block_id.get_slab_name(md5_shard));
+
+    ldpp_dout(dpp, 20) << __func__ << "::worker_id=" << (uint32_t)worker_id
+                       << ", md5_shard=" << (uint32_t)md5_shard
+                       << ", seq_number=" << seq_number
+                       << ":: oid=" << oid << dendl;
+
+    // read full object
+    int ret = ioctx.read(oid, bl, 0, 0);
+    if (ret > 0) {
+      ldpp_dout(dpp, 20) << __func__ << "::oid=" << oid << ", len=" << bl.length() << dendl;
+    }
+    else {
+      if (ret == 0) {
+        // no error reported, but we read nothing which should never happen
+        ldpp_dout(dpp, 5) << __func__ << "::ERR: Empty SLAB " << oid << dendl;
+        ret = -ENODATA;
```

Comment:
Since this is a "can't happen" condition should it have a higher log level (-1?) so if it does we'll get bug reports on it?

Reply:
okay